rpk: add client quotas support to rpk

r-vasquez opened the pull request:
This PR adds rpk support to DescribeClientQuotas and AlterClientQuotas under the `rpk cluster quotas` command.

### Examples:

**Creating a Quota**
```
$ rpk cluster quotas alter --add consumer_byte_rate=12000 --name client-id=foo --default user
ENTITY                        STATUS
client-id=foo,user=<default>  OK
```

**Describing Quotas**
```
$ rpk cluster quotas describe --name client-id=testclient                                    
ENTITY                               QUOTA-TYPE          QUOTA-VALUE
client-id=testclient,user=user2      producer_byte_rate  130000
client-id=testclient,user=<default>  producer_byte_rate  140000
```
All new commands now include support for `--format` json|yaml flag.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [X] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.1.x
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes

### Features

* rpk: now you can manage client quotas using `rpk cluster quotas`.

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/49661#018fc728-0975-4894-96e0-cf9cccd185bb
<!--fa417938-4afb-45a6-af04-cad6401c34c4-->

Deflaimun commented on src/go/rpk/pkg/cli/cluster/quotas/quotas.go:
```diff
@@ -0,0 +1,70 @@
+		Use:     "quotas",
+		Aliases: []string{"quota"},
+		Args:    cobra.NoArgs,
+		Short:   "Manage Redpanda client quotas",
```
question. Parent commands don't have long help texts?

r-vasquez replied:
Not necessarily, it depends on the complexity of the command and whether we want to explain a concept there.

r-vasquez commented:
Last Force Push:

- Refactor the `describe` output
- Add ducktape handler for `rpk cluster quotas` in 1b7fc43d8c523a4b39d4576a21d272a7f89f3190

pgellert commented on tests/rptest/clients/rpk.py:
```diff
@@ -1849,3 +1849,57 @@ def _run_role(self, cmd, output_format="json"):
 
         return json.loads(out) if output_format == "json" else out
+
+    def describe_cluster_quotas(self,
```
Thanks so much for setting this up!

Feediver1 commented on src/go/rpk/pkg/cli/cluster/quotas/alter.go:
```diff
@@ -0,0 +1,195 @@
+
+This command allows you to add or delete a client quota.
+
+A client quota consists of an entity (to whom the quota is applied) and a quota 
```
```suggestion
A client quota consists of an entity (to which the quota is applied) and a quota 
```

Feediver1 commented on src/go/rpk/pkg/cli/cluster/quotas/alter.go:
```diff
@@ -0,0 +1,195 @@
+Assigning quotas to default entity types is possible using the '--default' flag,
+a default user/client ID matches to all users or client IDs.
+
+A dry run can be performed using the '--dry' flag.
```
```suggestion
You can perform a dry run using the '--dry' flag.
```

Feediver1 commented on src/go/rpk/pkg/cli/cluster/quotas/alter.go:
```diff
@@ -0,0 +1,195 @@
+A dry run can be performed using the '--dry' flag.
+`,
+		Example: `
+Add quota (consumer_byte_rate) to client ID 'foo' and user 'bar'
```
```suggestion
Add quota (consumer_byte_rate) to client ID 'foo' and user 'bar':
```

Feediver1 commented on src/go/rpk/pkg/cli/cluster/quotas/alter.go:
```diff
@@ -0,0 +1,195 @@
+  rpk cluster quotas alter --add consumer_byte_rate=200000 \
+    --name client-id=foo,user=bar
+
+Add quota (producer_byte_rate) to client ID 'foo' and default users
```
```suggestion
Add quota (producer_byte_rate) to client ID 'foo' and default users:
```

Feediver1 commented on src/go/rpk/pkg/cli/cluster/quotas/alter.go:
```diff
@@ -0,0 +1,195 @@
+  rpk cluster quotas alter --add producer_byte_rate=180000 \
+    --name client-id=foo --default user
+
+Add quota (connection_creation_rate) to a given IP
```
```suggestion
Add quota (connection_creation_rate) to a given IP:
```

Feediver1 commented on src/go/rpk/pkg/cli/cluster/quotas/alter.go:
```diff
@@ -0,0 +1,195 @@
+  rpk cluster quotas alter --add connection_creation_rate=100 \
+    --name ip=181.81.18.81
+
+Remove quota (request_percentage) from client ID 'foo' and default users
```
```suggestion
Remove quota (request_percentage) from client ID 'foo' and default users:
```

Feediver1 commented on src/go/rpk/pkg/cli/cluster/quotas/alter.go:
```diff
@@ -0,0 +1,195 @@
+There are three entity types: client ID, user, or IP. You can assign quotas to 
+an entity consisting of a client ID and a user, or an IP.
+
+Assigning quotas to default entity types is possible using the '--default' flag,
+a default user/client ID matches to all users or client IDs.
```
```suggestion
Assigning quotas to default entity types is possible using the '--default' flag.
So if you want to assign a quota for a given client ID, but for _all_ users, then use this flag.
```

Feediver1 reviewed (dismissed):
Left some suggestions for clarity, consistency with docs.

Feediver1 commented on src/go/rpk/pkg/cli/cluster/quotas/describe.go:
```diff
@@ -0,0 +1,173 @@
+		Long: `Describe client quotas.
+
+This command describes client quotas that match the provided filtering criteria.
+Running the command without filters will return all client quotas. The 
+'--strict' flag can be used for strict matching (only returning quotas that 
+exactly match the filters).
+
+Filters can be provided in terms of entities. An entity consists of either a 
+client ID and user pair or an IP address.
```
```suggestion
Running the command without filters returns all client quotas. Use the 
'--strict' flag for strict matching, which means that the only quotas returned 
exactly match the filters).

Filters can be provided in terms of entities. An entity consists of either a 
client ID and user pair or an IP address.
```
```suggestion
Running the command without filters will return all client quotas. The 
'--strict' flag can be used for strict matching (only returning quotas that 
exactly match the filters).

Filters can be provided in terms of entities. An entity consists of either a 
client ID and user pair or an IP address.
```

r-vasquez commented:
@pgellert How does a client-id-prefix entity work along with client-id? how does a default for client-id-prefix work?

Is an entity composed of `{client-id=id, client-id-prefix=prefix}` or is it like IP, where the prefix is a single entity.

Also, out of curiosity, what's the order of precedence? say I have:
```
client-id=foo-bar
consumer_byte_rate=200

client-id=<default>
consumer_byte_rate=100

client-id-prefix=foo-
consumer_byte_rate=300
```

pgellert commented:
@r-vasquez:
> How does a client-id-prefix entity work along with client-id? how does a default for client-id-prefix work?

The priority order is:
1. Explicit match: `(client-id, <client-id>)`
2. Prefix match: `(client-id-prefix, <client-id-prefix>)`
3. Client default: `(client-id, <default>)`

There's no specific default for `client-id-prefix`, `(client-id-prefix, <default>)` should be "not supported" and we should error out either in rpk/redpanda.

> Is an entity composed of `{client-id=id, client-id-prefix=prefix}` or is it like IP, where the prefix is a single entity.

The `client-id` and `client-id-prefix` should be mutually exclusive, but we should allow both `{client-id=id, user=user}` and `{client-id-prefix=prefix, user=user}` pairs (at least once redpanda supports user-based quotas). So `{client-id-prefix=prefix}` should be usable whereever `{client-id=id}` is used.

> Also, out of curiosity, what's the order of precedence? say I have:
> 
> ```
> client-id=foo-bar
> consumer_byte_rate=200
> 
> client-id=<default>
> consumer_byte_rate=100
> 
> client-id-prefix=foo-
> consumer_byte_rate=300
> ```

Based on the priority order from above, `foo-bar` would get 200 (explicit match wins over prefix match), `foo-abc` would get 300 (from the prefix match) and `baz` would get 100 (from the default).

r-vasquez commented:
Last 2 force pushes: [1](https://github.com/redpanda-data/redpanda/compare/1b7fc43d8c523a4b39d4576a21d272a7f89f3190..6db00dbe2dd92174d29b88cf4af2f4c6b22c616a) and [2](https://github.com/redpanda-data/redpanda/compare/6db00dbe2dd92174d29b88cf4af2f4c6b22c616a..9e4f61ed45d0e78dd629652cfe593a74f21d5d80)
- Removed any reference to `user` and `ip` from the help text.
- Added `client-id-prefix` to the examples.
- Added an error message when the `describe` command returns an empty list. (text format)

@pgellert Thanks for the context, I changed the PR to meet what Redpanda will support, the validation happens in Redpanda so we should be fine.

(For example, this is using Kafka:)
```
// client-id-prefix is not valid in Kafka
$ rpk cluster quotas alter --add producer_byte_rate=12123123 --name client-id-prefix=foo-
ENTITY                 STATUS
client-id-prefix=foo-  Error: Unhandled client quota entity type: client-id-prefix

// From Kafka Logs:
[2024-06-05 15:21:02,385] INFO [Admin Manager on Broker 1]: Error encountered while updating client quotas (kafka.server.ZkAdminManager)
org.apache.kafka.common.errors.InvalidRequestException: Unhandled client quota entity type: client-id-prefix
```

Please review the examples again.

pgellert commented on src/go/rpk/pkg/cli/cluster/quotas/describe.go:
```diff
@@ -0,0 +1,174 @@
+					MatchType: 0,
+				})
+			}
+			for _, def := range defaults {
+				k := strings.ToLower(def)
+				if !validTypes[k] {
+					out.Die("default type %q is invalid (allowed: client-id, client-id-prefix)", def)
+				}
+				reqQuotas = append(reqQuotas, kadm.DescribeClientQuotaComponent{
+					Type:      k,
+					MatchType: 1,
+				})
+			}
```
Same point about `--default client-id-prefix` here.

pgellert commented on src/go/rpk/pkg/cli/cluster/quotas/alter.go:
```diff
@@ -0,0 +1,194 @@
+				})
+			}
+			for _, def := range defaults {
+				if !validTypes[def] {
+					out.Die("default type %q is invalid (allowed: client-id, client-id-prefix)", def)
+				}
```
The error message says that `--default client-id-prefix` is allowed even though it's not supported. I think there should probably be a separate `validTypes` for `any` and `default`. Alternatively, we could handle the validation in redpanda and rpk will just show the error but at least this error message should be changed.

r-vasquez replied:
I added your suggestion in a new commit, thanks! https://github.com/redpanda-data/redpanda/pull/18711/commits/0ac43ee9a4a6c22c7614a1e823dce9128e0652b5

pgellert reviewed (commented):
There's one error message which is misleading but apart from that lgtm.
